Best of last year

aehyvari opened the pull request:
This PR provides more automation for extracting the previous year's best solvers and constructing the corresponding jobs for running them.  In addition it contains an updated `solvers_divisions_final.csv`.

- `score.py`
    - The script `score.py` flag --bestof takes as argument the csv file where the best of last year solvers will be printed
    - The output will contain the last year best solver
- `tools/previous-best/best_solvers_to_csv.py`
    - A new script that converts the csv file from score.py into a form compatible with this years `solvers_divisions_final.csv`
- `2020/previous-best/get_last_year_best_solvers.sh`
    - A new script that puts together the pipeline consisting of `score.py` and `best_solvers_to_csv.py` and produces the excerpts ready to be added to `solvers_divisions_final.csv`
- `tools/prep/prepare_space_xml.py`
    - Enabled the forcing of running non-competitive divisions.  Useful if re-running jobs based on a stripped solvers_divisions.csv
- `2020/prep/submitted/make_spaces.sh`
    - Added an option `--solvdiv` that takes as argument an alternate `solvers_divisions.csv`.  When given, uses the provided csv and enables running of non-competitive divisions.  Useful for re-running jobs.

HanielB commented on tools/scoring/score.py:
```diff
@@ -1882,10 +1882,21 @@ def main():
-              print("Best competitive solvers:\n" + winners[["name"]].to_csv())
+              winners = grouped[\
+                          (grouped["rank"] == 1) &\
+                          (grouped["score_correct"] > 0) &\
+                          (grouped["score_error"] == 0)
```
What is the motivation for this? Divisions where all solvers were unsound?

aehyvari replied:
Yes.  I don't think there are any, but I thought it's a good way of spotting them.

aehyvari replied:
Inside the rank == 1 the solvers are ordered by the other criteria of the scoring system (the CPU/wall-clock time mainly, at least in the single-query track).  So the first of those solvers that has rank == 1 is the winner.

HanielB commented on tools/scoring/score.py:
```diff
@@ -1882,10 +1882,21 @@ def main():
+                          (grouped["score_error"] == 0)
+                        ]
+              winners = winners.reset_index()\
+                .drop_duplicates(["year", "division"], keep="first")
```
why drop_duplicates?

aehyvari replied:
Inside the rank == 1 the solvers are ordered by the other criteria of the scoring system (the CPU/wall-clock time mainly, at least in the single-query track). So the first of those solvers that has rank == 1 is the winner.

HanielB commented on tools/previous-best/README.md:
```diff
@@ -0,0 +1,21 @@
+## Previous year's winners
+
+Given the best solvers from last year, `best_solvers_to_csv.py` search them from the
```
```suggestion
Given the best solvers from last year, `best_solvers_to_csv.py` searches for them from the
```

HanielB commented on tools/previous-best/README.md:
```diff
@@ -0,0 +1,21 @@
+ - best-solvers: the best solvers csv with headers for division, and
+   solver name, for each track.
+
+The script only outputs the new lines, they are not automatically added
+to `solvers_divisions.csv`.
```
As I mentioned, I think that it'd be best to have this script transform the given `new-registration` csv by adding the entries of the bestof solvers.

aehyvari replied:
I didn't want to automate that, since something unexpected could mess up the file.  When touching that file I think some human intervention would be good.  But I'm open to ideas: for instance the changes will have to be committed anyway, and maybe that's a good enough point to do it.  But at the same time that can slip past the eyes of a tired committer...

HanielB commented on tools/prep/prepare_space_xml.py:
```diff
@@ -255,7 +255,7 @@ def add_solvers():
                     # Consequently, solvers do not get added to non-competitive
                     # divisions, which are then removed below.
-                    if is_competitive(solvers):
+                    if g_args.run_noncompetitive or is_competitive(solvers):
```
this is redundant

aehyvari replied:
I agree.

HanielB commented on tools/prep/prepare_space_xml.py:
```diff
@@ -316,6 +316,7 @@ def main():
                              "solvers (StarExec IDs, wrapped IDs if -w)")
     parser.add_argument("--include-non-competitive",
                         default=False,
+                        dest="run_noncompetitive",
```
why add this?

aehyvari replied:
I just find it nicer to write and remember.  Do you prefer the include_non_competitive?  Or did you mean this question more generally?

jhoenicke commented on 2020/prep/submitted/make_spaces.sh:
```diff
@@ -111,20 +115,28 @@ done
-python $PREPARE "$IN_SPACE_NON_INC" "$SOLVERS_CSV" "$OUT_SPACE_SQ" -t single_query --select "$SELECT_SQ" -w
+python $PREPARE "$IN_SPACE_NON_INC" "$SOLVERS_CSV" "$OUT_SPACE_SQ" \
+    -t single_query --select "$SELECT_SQ" -w \
+    $(echo $INCLUDE_NONCOMPETITIVE)
```
The `echo` is unnecessary:
```suggestion
    $INCLUDE_NONCOMPETITIVE
```

Just be sure to not include quotes, so that no argument is added when the variable is emty.

jhoenicke commented on 2020/previous-best/get_last_year_best_solvers.sh:
```diff
@@ -0,0 +1,69 @@
+UNSAT_CORE_RESULTS_2019="../../2019/results/Unsat_Core_Track.csv"
+MODEL_VALIDATION_RESULTS_2019="../../2019/results/Model_Validation_Track.csv"
+
+INCREMENTAL_NUM_CHECKSATS_2019="./num_check_sats_2019.csv"
```
```suggestion
INCREMENTAL_NUM_CHECKSATS_2019="../../2019/prep/incremental_num_check_sat.csv"
```

and then you can remove the tar.xz file.

jhoenicke commented on tools/previous-best/best_solvers_to_csv.py:
```diff
@@ -0,0 +1,154 @@
+                        new_winner_row.append(adjustName(name, \
+                            g_args.old_year))
+                    else:
+                        new_winner_row.append(quoteIfHasComma(row[col]))
```
Should we remove the contents of all columns except for the wrapped solver ids?

aehyvari replied:
Also @HanielB brought this up.  My justification of keeping them is that maybe an occasional browser of the website would like to see this information.  Pointing out by these empty fields that this is not an actual submission seems a little too subtle anyway.

HanielB commented:
Indeed the current make_spaces.sh is broken. Using a single CSV the fix is simple though (which is the same it was used last year): add ```--exclude-solvers "$FIXED_SOLVERS_NON_COMPETITIVE_SQ,$SOLVERS_2018_SQ"```. It does not look to me much more complicated than having one CSV for competition, one for bestof, one for hot-fixed. A single CSV as a reference point for the ran solvers sounds easier to manage to me.

aehyvari commented:
> A more general one on the changes to the solvers final CSV: It was by design that the bestof entries had no other information other than the IDs, the solver name and the non-competitive bit. I think it makes sense with the fact that these are entries that we, the organizers, are adding, not the people that submitted in the previous year.

I think it's good to have this information available for people who go to the website and want more information on the best solvers.  If these get confused with actually competing solvers, we could come up with less subtle ways of clarifying this.

> Another thing: the logics that you have in the solvers CSV have changed from when I added. Why is that?

The only change should be single-query AUFNIA, which I removed because [it didn't have a winner last year](https://smt-comp.github.io/2019/results/aufnia-single-query).

aehyvari commented:
> Indeed the current make_spaces.sh is broken. Using a single CSV the fix is simple though (which is the same it was used last year): add --exclude-solvers "$FIXED_SOLVERS_NON_COMPETITIVE_SQ,$SOLVERS_2018_SQ". It does not look to me much more complicated than having one CSV for competition, one for bestof, one for hot-fixed. A single CSV as a reference point for the ran solvers sounds easier to manage to me.

I changed the approach a little.  Now there is a single csv, but I added two scripts, [`make_spaces_best_of_2019.sh`](https://github.com/SMT-COMP/smt-comp/blob/best-of-last-year/2020/prep/submitted/make_spaces_best_of_2019.sh), and [`make_spaces_main.sh`](https://github.com/SMT-COMP/smt-comp/blob/best-of-last-year/2020/prep/submitted/make_spaces_main.sh) that run `make_spaces.sh` with a temporary csv file.  The temporary csv file relies on a naming convention that the previous year's best solvers are prepended with `2018-` or `2019-`.

I hope this fixes both of the concerns, having to maintain separate csv files and having no way of constructing only the main jobs without editing the scripts.
